src/crypto/ocb_openssl.cc

```diff
@@ -20,49 +23,63 @@ int ae_ctx_sizeof(void) {
-  // Pre-condition: Only nonces of length 12 are supported. The
-  // documentation of `ae_init` in ae.h specifies that `ctx` is
-  // untouched if an invalid configuration is requested. Delegating
-  // this to OpenSSL would happen too late; `ctx` has already been
-  // modified.
-  if (nonce_len != 12) {
-    return AE_NOT_SUPPORTED;
-  }
-
-  // Pre-condition: Only AES-128 is supported.
-  if (key_len != 16) {
-    return AE_NOT_SUPPORTED;
-  }
-
+// If direction is 1, initializes encryption. If 0, initializes
+// decryption. See the documentation of EVP_CipherInit_ex
+static int ae_evp_cipher_init(EVP_CIPHER_CTX **in_ctx, int direction,
+			      int nonce_len, int tag_len) {
   // Create an OpenSSL EVP context. It does not yet have any specific
   // cipher associated with it.
-  if (!(ctx->ocb_ctx = EVP_CIPHER_CTX_new())) {
+  if (!(*in_ctx = EVP_CIPHER_CTX_new())) {
     return -10;
   }
-  // ae_init is used for both OCB encryption and decryption. However,
-  // the EVP_CIPHER_CTX wants to know if the operation is encryption
-  // or decryption. For now, arbitrary pick encryption. This will get
-  // overriden in the `ae_encrypt` and `ae_decrypt` methods when the
-  // key and nonce (IV) are loaded.
-  if (EVP_EncryptInit_ex(ctx->ocb_ctx, EVP_aes_128_ocb(),
-			 NULL, NULL, NULL) != 1) {
+  EVP_CIPHER_CTX *ctx = *in_ctx;
+  // Initialize either encryption or decryption, based on the argument
+  // of `direction` passed in. Although OCB-AES doesn't really believe
+  // in different operations for encryption and decryption, OpenSSL
+  // does.
+  if (EVP_CipherInit_ex(ctx, EVP_aes_128_ocb(),
+			NULL, NULL, NULL, direction) != 1) {
     return -11;
   }
   // Attempt to set the nonce length. If it fails, the length must not
   // be supported. However, that should have been handled by the
   // pre-condition check above.
-  if (EVP_CIPHER_CTX_ctrl(ctx->ocb_ctx, EVP_CTRL_AEAD_SET_IVLEN,
+  if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN,
 			  nonce_len, NULL)!= 1) {
     return -12;
   }
   // Attempt to set the tag length. A NULL tag length means that
-  // EVP_CRTL_AEAD_SET_TAG is only being used to set the length
-  if (EVP_CIPHER_CTX_ctrl(ctx->ocb_ctx, EVP_CTRL_AEAD_SET_TAG,
+  // EVP_CTRL_AEAD_SET_TAG is only being used to set the length
+  if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG,
 			  tag_len, NULL) != 1) {
     return -13;
   }
+  return AE_SUCCESS;
+}
+
+int ae_init(ae_ctx *ctx, const void *key_param, int key_len, int nonce_len,
+	    int tag_len) {
+  // Pre-condition: Only nonces of length 12 are supported. The
+  // documentation of `ae_init` in ae.h specifies that `ctx` is
+  // untouched if an invalid configuration is requested. Delegating
+  // this to OpenSSL would happen too late; `ctx` has already been
+  // modified.
+  if (nonce_len != 12) {
+    return AE_NOT_SUPPORTED;
+  }
+  // Pre-condition: Only AES-128 is supported.
+  if (key_len != 16) {
+    return AE_NOT_SUPPORTED;
+  }
+  int r = 0;
```

Comment:
Very nitpicky nitpick: It's a little odd to initialize this to zero when the type normally has constants (zero is `AE_SUCCESS`). I'd probably have omitted it. `AE_SUCCESS` seems also defensible. (In the steady state, `r` is `AE_SUCCESS` because any failure should have returned out.) Or some error in an attempt to fail closed. Not that it matters in the slightest.